components/tidb_query/src/codec/mysql/json/binary.rs

```diff
@@ -183,18 +259,18 @@ pub trait JsonEncoder: NumberEncoder {
         Ok(())
     }
 
-    fn write_json_item(
+    fn write_json_item<'a>(
         &mut self,
-        data: &Json,
+        data: &JsonRef<'a>,
         offset: &mut u32,
         data_buf: &mut Vec<u8>,
     ) -> Result<()> {
-        let code = data.get_type_code();
-        self.write_u8(code)?;
-        match *data {
+        let code = data.get_type();
+        self.write_u8(code as u8)?;
+        match code {
             // If the data has length in (0, 4], it could be inline here.
             // And padding 0x00 to 4 bytes if needed.
-            Json::Boolean(_) | Json::None => {
+            JsonType::Literal => {
```

Comment:
Seems the logic is different from TiDB. 
See: https://github.com/pingcap/tidb/blob/master/types/json/binary.go#L569:L569, there is no padding for literal.

Reply:
The `buf` is pre-allocated and each value entry must be 5 bytes long